core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -261,7 +269,27 @@ private[spark] class CoarseMesosSchedulerBackend(
     val taskId = status.getTaskId.getValue.toInt
     val state = status.getState
     logInfo(s"Mesos task $taskId is now $state")
+    val slaveId: String = status.getSlaveId.getValue
     stateLock.synchronized {
+      if (TaskState.fromMesos(state).equals(TaskState.RUNNING) &&
+        slaveIdsToShuffleEndpoints.contains(slaveId)) {
+        val rpcAddress = slaveIdsToShuffleEndpoints.remove(slaveId).get
+        logDebug(s"Connecting to shuffle service on slave ${slaveId}, " +
+            s"address $rpcAddress for app ${conf.getAppId}")
+        sc.env.rpcEnv.asyncSetupEndpointRef(
+          MesosExternalShuffleService.SYSTEM_NAME,
+          rpcAddress,
+          MesosExternalShuffleService.ENDPOINT_NAME).onComplete { ref =>
+          if (ref.isSuccess) {
+            ref.get.send(RegisterMesosDriver(conf.getAppId, sc.env.rpcEnv.address))
```

Comment:
also nit: you use `sc.env.rpcEnv` in a few places, can you make it a local variable `val rpcEnv`